cornac/data/reader.py

```diff
@@ -234,18 +279,13 @@ def read(
         """
         parser = PARSERS.get(fmt, None) if parser is None else parser
         if parser is None:
-            raise ValueError(
-                "Invalid line format: {}\n"
-                "Only support: {}".format(fmt, PARSERS.keys())
-            )
+            raise ValueError("Invalid line format: {}\n" "Only support: {}".format(fmt, PARSERS.keys()))
 
         with open(fpath, encoding=self.encoding, errors=self.errors) as f:
             tuples = [
                 tup
                 for idx, line in enumerate(itertools.islice(f, skip_lines, None))
-                for tup in parser(
-                    line.strip().split(sep), line_idx=idx, id_inline=id_inline, **kwargs
-                )
+                for tup in parser(line.strip().split(sep), line_idx=idx, id_inline=id_inline, **kwargs)
             ]
             return self._filter(tuples)
```

Comment:
Consider adding two functions `_filter_sequence()` and `_filter_basket()` and only apply them for corresponding data format. This will make the filtering steps easier to manage.